### PR TITLE
Fix `apply` performance

### PR DESCRIPTION
**Pull request by rafaqz**
This should mostly fix `tuples` and `flip` performance. Not perfect yet but removes 95% of the type stability problems.

### CODE REVIEW

File: test/transformations/flip.jl [hunk 1]
```diff
@@ -4,10 +4,12 @@ import GeoInterface as GI
 import GeometryOps as GO
 
 @testset "flip" begin
+    using ProfileView, Cthulhu
```
Comment (asinghvi17) [inline test/transformations/flip.jl:7]:
```suggestion
    # using ProfileView, Cthulhu

File: test/transformations/flip.jl [hunk 2]
```diff
@@ -4,10 +4,12 @@ import GeoInterface as GI
...
     geom = GI.Polygon([GI.LinearRing([(1, 2), (3, 4), (5, 6), (1, 2)]), 
                        GI.LinearRing([(3, 4), (5, 6), (6, 7), (3, 4)])])
 
 
     @test GO.flip(geom) == GI.Polygon([GI.LinearRing([(2, 1), (4, 3), (6, 5), (2, 1)]), 
                                        GI.LinearRing([(4, 3), (6, 5), (7, 6), (4, 3)])])
+    @profview for i in 1:10000 GO.flip(geom) end
```
Comment (asinghvi17) [inline test/transformations/flip.jl:14]:
```suggestion
    # @profview for i in 1:10000 GO.flip(geom) end

File: test/transformations/flip.jl (hunk 1 above)
Comment (rafaqz) [reply test/transformations/flip.jl:7]: lol

Comment (rafaqz) [reply test/transformations/flip.jl:7]: sharing all my trade secrets

File: src/primitives.jl [hunk 3]
```diff
@@ -89,42 +89,48 @@ flipped_geom = GO.apply(GI.PointTrait, geom) do p
     (GI.y(p), GI.x(p))
 end
 """
-apply(f, ::Type{Target}, geom; kw...) where Target = _apply(f, Target, geom; kw...)
+Base.@constprop :aggressive apply(f::F, ::Type{Target}, geom; threaded=false, kw...) where {F,Target} =
+    _apply(f, Target, geom; threaded=_threadtype(threaded), kw...)
+
```
Comment (asinghvi17) [inline src/primitives.jl:94]:
```suggestion

#=
You might note the use of this strange macro, `Base.@constprop :aggressive`.  This tells Julia that the return type of the method depends on the value of its input arguments, and allows it to specialize compilation to increase type-stability.

We are also passing the threading specification as types, not simple boolean values.  This is, again, to help compilation - with a type to hold on to, it's easier for the compiler to separate threaded and non-threaded code paths.
=#

```

Comment (rafaqz) [reply src/primitives.jl:94]: I should go through and write this stuff for everything

File: src/primitives.jl [hunk 4]
```diff
@@ -89,42 +89,54 @@ flipped_geom = GO.apply(GI.PointTrait, geom) do p
...
     # Run _apply on all `features` in the feature collection, possibly threaded
-    features = _maptasks(1:GI.nfeature(fc); threaded) do i
-        feature = GI.getfeature(fc, i)
-        _apply(f, Target, feature; crs, calc_extent, threaded=false)::GI.Feature
-    end
-    if calc_extent
+    apply_to_feature(i) =
+        _apply(f, Target, GI.getfeature(fc, i); crs, calc_extent, threaded=_False())::GI.Feature
+    features = _maptasks(apply_to_feaature, 1:GI.nfeature(fc), threaded)
```
Comment (asinghvi17) [inline src/primitives.jl:138]: Typo here

Comment (asinghvi17) [reply src/primitives.jl:138]:
```suggestion
    features = _maptasks(apply_to_feature, 1:GI.nfeature(fc), threaded)

**Pull request merged**